api/admin/controller.py

```diff
@@ -2240,42 +2330,104 @@ def admin_auth_service(self, protocol):
     def individual_admins(self):
         if flask.request.method == 'GET':
             admins = []
-            admins_with_password = Admin.with_password(self._db)
-            if admins_with_password.count() != 0:
-                admins=[dict(email=admin.email) for admin in admins_with_password]
+            for admin in self._db.query(Admin):
+                roles = []
+                for role in admin.roles:
+                    if role.library:
+                        if not flask.request.admin or not flask.request.admin.is_librarian(role.library):
+                            continue
+                        roles.append(dict(role=role.role, library=role.library.short_name))
+                    else:
+                        roles.append(dict(role=role.role))
+                admins.append(dict(email=admin.email, roles=roles))
 
             return dict(
                 individualAdmins=admins,
             )
 
         email = flask.request.form.get("email")
         password = flask.request.form.get("password")
+        roles = flask.request.form.get("roles")
 
-        if not email or not password:
+        if not email:
             return INCOMPLETE_CONFIGURATION
 
+        if roles:
+            roles = json.loads(roles)
+        else:
+            roles = []
+
         admin, is_new = get_one_or_create(self._db, Admin, email=email)
-        admin.password = password
+        if admin.is_sitewide_librarian():
+            self.require_sitewide_library_manager()
+        if admin.is_system_admin():
+            self.require_system_admin()
+            
+        if password:
+            admin.password = password
         try:
             self._db.flush()
         except ProgrammingError as e:
             self._db.rollback()
             return MISSING_PGCRYPTO_EXTENSION
 
+        old_roles = admin.roles
+        for role in roles:
+            if role.get("role") not in AdminRole.ROLES:
+                self._db.rollback()
+                return UNKNOWN_ROLE
+            library = None
+            library_short_name = role.get("library")
+            if library_short_name:
+                library = Library.lookup(self._db, library_short_name)
+                if not library:
+                    self._db.rollback()
+                    return LIBRARY_NOT_FOUND.detailed(_("Library \"%(short_name)s\" does not exist.", short_name=library_short_name))
+            if library:
+                self.require_library_manager(library)
+            elif role.get("role") == AdminRole.SYSTEM_ADMIN:
+                self.require_system_admin()
+            else:
+                self.require_sitewide_library_manager()
+            admin.add_role(role.get("role"), library)
```

Comment:
I have a lot of logic on the client to adjust all the roles as things are added or removed. For example, if you are a librarian of a few libraries and you become a sitewide librarian, it removes the individual librarian roles. I wasn't sure if it was necessary to handle that on the server as well. If everyone's using the client to set roles it won't be necessary.

Reply:
I think it's not necessary to handle that on the server since it's not inconsistent for you to have a few libraries + a sitewide role, just redundant.

Reply:
It can be inconsistent if you go the other direction, like if you are a sitewide librarian and removed from one library. On the client removing you from a library when you have a sitewide role will remove the sitewide role and create individual roles for every other library.

Reply:
It doesn't leave an inconsistent state in the database, but it could be inconsistent with users' expectations.

Reply:
I'm OK with that translation happening entirely on the client side. If a request somehow goes in to remove a sitewide admin from one specific library, I would expect that to be a no-op.